File: snippets/screens/swiftui/search_history_screen/2/search_history_screen.swift
import SwiftUI

struct SearchHistoryItem {
    let id: String
    let query: String
    let timestamp: Date
    let resultCount: Int
    let category: String?
    let filters: [String]
}

struct SearchHistoryScreen: View {
    @State private var searchHistory: [SearchHistoryItem] = []
    @State private var filteredHistory: [SearchHistoryItem] = []
    @State private var searchText = ""
    @State private var isLoading = true
    @State private var selectedTimeFilter = "all"
    @State private var isSelectionMode = false
    @State private var selectedItems = Set<String>()

    private let timeFilters = [
        ("all", "All time"),
        ("today", "Today"),
        ("week", "This week"),
        ("month", "This month")
    ]

    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Search Bar
                HStack {
                    TextField("Search in history...", text: $searchText)
                        .textFieldStyle(.roundedBorder)
                        .onChange(of: searchText) { _, _ in
                            filterHistory()
                        }

                    if !searchText.isEmpty {
                        Button {
                            searchText = ""
                            filterHistory()
                        } label: {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundColor(.gray)
                        }
                    }
                }
                .padding(.horizontal)

                // Time Filter Chips
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 12) {
                        ForEach(timeFilters, id: \.0) { filter in
                            timeFilterChip(filter.1, value: filter.0)
                        }
                    }
                    .padding(.horizontal)
                }
                .padding(.vertical, 8)

                // Content
                if isLoading {
                    Spacer()
                    ProgressView()
                    Spacer()
                } else if filteredHistory.isEmpty {
                    emptyStateView
                } else {
                    historyList
                }
            }
            .navigationTitle(isSelectionMode ? "\(selectedItems.count) selected" : "Search History")
            .navigationBarTitleDisplayMode(.large)
            .toolbar {
                ToolbarItemGroup(placement: .navigationBarLeading) {
                    if isSelectionMode {
                        Button("Cancel") {
                            isSelectionMode = false
                            selectedItems.removeAll()
                        }
                    }
                }

                ToolbarItemGroup(placement: .navigationBarTrailing) {
                    if isSelectionMode {
                        if !selectedItems.isEmpty {
                            Button {
                                deleteSelectedItems()
                            } label: {
                                Image(systemName: "trash")
                            }
                        }
                    } else {
                        Button {
                            isSelectionMode = true
                        } label: {
                            Image(systemName: "checkmark.circle")
                        }

                        Menu {
                            Button("Clear All History", role: .destructive) {
                                clearAllHistory()
                            }
                        } label: {
                            Image(systemName: "ellipsis.circle")
                        }
                    }
                }
            }
        }
        .onAppear {
            loadSearchHistory()
        }
    }

    private func timeFilterChip(_ label: String, value: String) -> some View {
        let isSelected = selectedTimeFilter == value

        return Button {
            selectedTimeFilter = value
            filterHistory()
        } label: {
            Text(label)
                .font(.caption)
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(isSelected ? Color.accentColor : Color(.systemGray6))
                .foregroundColor(isSelected ? .white : .primary)
                .cornerRadius(16)
        }
        .buttonStyle(.plain)
    }

    private var emptyStateView: some View {
        VStack(spacing: 16) {
            Spacer()

            Image(systemName: "clock")
                .font(.system(size: 64))
                .foregroundColor(.gray)

            Text(searchText.isEmpty ? "No search history" : "No matching searches found")
                .font(.title2)
                .fontWeight(.bold)

            Text(searchText.isEmpty ? "Your recent searches will appear here" : "Try different keywords")
                .font(.body)
                .foregroundColor(.secondary)

            Spacer()
        }
    }

    private var historyList: some View {
        List {
            ForEach(groupedHistory.keys.sorted { groupOrder($0) < groupOrder($1) }, id: \.self) { group in
                Section {
                    ForEach(groupedHistory[group] ?? [], id: \.id) { item in
                        historyItem(item)
                    }
                } header: {
                    Text(group)
                        .font(.subheadline)
                        .fontWeight(.bold)
                        .foregroundColor(.accentColor)
                }
            }
        }
        .listStyle(.plain)
    }

    private var groupedHistory: [String: [SearchHistoryItem]] {
        let calendar = Calendar.current
        let now = Date()

        return Dictionary(grouping: filteredHistory) { item in
            if calendar.isDateInToday(item.timestamp) {
                return "Today"
            } else if calendar.isDateInYesterday(item.timestamp) {
                return "Yesterday"
            } else if item.timestamp > calendar.date(byAdding: .weekOfYear, value: -1, to: now) ?? now {
                return "This week"
            } else if item.timestamp > calendar.date(byAdding: .month, value: -1, to: now) ?? now {
                return "This month"
            } else {
                return "Older"
            }
        }
    }

    private func groupOrder(_ group: String) -> Int {
        switch group {
        case "Today": return 0
        case "Yesterday": return 1
        case "This week": return 2
        case "This month": return 3
        case "Older": return 4
        default: return 5
        }
    }

    private func historyItem(_ item: SearchHistoryItem) -> some View {
        HStack {
            if isSelectionMode {
                Button {
                    toggleItemSelection(item.id)
                } label: {
                    Image(systemName: selectedItems.contains(item.id) ? "checkmark.circle.fill" : "circle")
                        .foregroundColor(selectedItems.contains(item.id) ? .accentColor : .gray)
                }
                .buttonStyle(.plain)
            } else {
                Image(systemName: "clock")
                    .foregroundColor(.accentColor)
                    .frame(width: 24)
            }

            VStack(alignment: .leading, spacing: 4) {
                Text(item.query)
                    .font(.body)
                    .fontWeight(.medium)

                HStack(spacing: 8) {
                    Text(relativeTime(from: item.timestamp))
                        .font(.caption)
                        .foregroundColor(.secondary)

                    if let category = item.category {
                        Text(category)
                            .font(.caption)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color.accentColor.opacity(0.1))
                            .foregroundColor(.accentColor)
                            .cornerRadius(4)
                    }

                    ForEach(item.filters, id: \.self) { filter in
                        Text(filter)
                            .font(.caption)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color(.systemGray5))
                            .foregroundColor(.secondary)
                            .cornerRadius(4)
                    }
                }
            }

            Spacer()

            if !isSelectionMode {
                Menu {
                    Button {
                        UIPasteboard.general.string = item.query
                    } label: {
                        Label("Copy", systemImage: "doc.on.clipboard")
                    }

                    Button("Delete", role: .destructive) {
                        deleteHistoryItem(item.id)
                    }
                } label: {
                    Image(systemName: "ellipsis")
                        .foregroundColor(.gray)
                }
            }
        }
        .contentShape(Rectangle())
        .onTapGesture {
            if isSelectionMode {
                toggleItemSelection(item.id)
            } else {
                performSearch(item.query)
            }
        }
    }

    private func loadSearchHistory() {
        isLoading = true

        // Simulate loading from storage
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            searchHistory = generateMockHistory()
            filteredHistory = searchHistory
            isLoading = false
        }
    }

    private func generateMockHistory() -> [SearchHistoryItem] {
        let queries = [
            "Swift programming",
            "iOS development",
            "SwiftUI tutorials",
            "Xcode tips",
            "Core Data",
            "Combine framework",
            "App Store optimization",
            "TestFlight beta",
            "Apple HIG",
            "Performance optimization",
            "Memory management",
            "Auto Layout",
            "Dark mode",
            "Accessibility",
            "Push notifications"
        ]

        let categories = ["Development", "Design", "Testing", "Distribution"]

        return queries.enumerated().map { index, query in
            let daysAgo = index * 2 + (index % 3)
            return SearchHistoryItem(
                id: "search_\(index)",
                query: query,
                timestamp: Calendar.current.date(byAdding: .day, value: -daysAgo, to: Date()) ?? Date(),
                resultCount: 50 + (index * 15),
                category: categories[index % categories.count],
                filters: index % 3 == 0 ? ["iOS", "Popular"] : index % 4 == 0 ? ["Recent", "Trending"] : []
            )
        }
    }

    private func filterHistory() {
        filteredHistory = searchHistory.filter { item in
            let matchesSearch = searchText.isEmpty || item.query.localizedCaseInsensitiveContains(searchText)
            let matchesTimeFilter = matchesTimeFilter(item)
            return matchesSearch && matchesTimeFilter
        }
    }

    private func matchesTimeFilter(_ item: SearchHistoryItem) -> Bool {
        let calendar = Calendar.current
        let now = Date()

        switch selectedTimeFilter {
        case "today":
            return calendar.isDateInToday(item.timestamp)
        case "week":
            return item.timestamp > calendar.date(byAdding: .weekOfYear, value: -1, to: now) ?? now
        case "month":
            return item.timestamp > calendar.date(byAdding: .month, value: -1, to: now) ?? now
        default:
            return true
        }
    }

    private func performSearch(_ query: String) {
        // Navigate to search results with query
        // This would typically use NavigationLink or programmatic navigation
    }

    private func deleteHistoryItem(_ id: String) {
        searchHistory.removeAll { $0.id == id }
        filteredHistory.removeAll { $0.id == id }
        // In a real app, you would also remove from persistent storage
    }

    private func deleteSelectedItems() {
        searchHistory.removeAll { selectedItems.contains($0.id) }
        filteredHistory.removeAll { selectedItems.contains($0.id) }
        selectedItems.removeAll()
        isSelectionMode = false
        // In a real app, you would also remove from persistent storage
    }

    private func clearAllHistory() {
        searchHistory.removeAll()
        filteredHistory.removeAll()
        // In a real app, you would also clear persistent storage
    }

    private func toggleItemSelection(_ id: String) {
        if selectedItems.contains(id) {
            selectedItems.remove(id)
        } else {
            selectedItems.insert(id)
        }
    }

    private func relativeTime(from date: Date) -> String {
        let now = Date()
        let calendar = Calendar.current
        let components = calendar.dateComponents([.minute, .hour, .day, .weekOfYear], from: date, to: now)

        if let weeks = components.weekOfYear, weeks > 0 {
            let formatter = DateFormatter()
            formatter.dateStyle = .short
            return formatter.string(from: date)
        } else if let days = components.day, days > 0 {
            return days == 1 ? "1 day ago" : "\(days) days ago"
        } else if let hours = components.hour, hours > 0 {
            return hours == 1 ? "1 hour ago" : "\(hours) hours ago"
        } else if let minutes = components.minute, minutes > 0 {
            return minutes == 1 ? "1 minute ago" : "\(minutes) minutes ago"
        } else {
            return "Just now"
        }
    }
}

#Preview {
    SearchHistoryScreen()
}

File: snippets/screens/flutter/search_history_screen/3/search_history_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class SearchHistoryItem {
  final String id;
  final String query;
  final DateTime timestamp;
  final int resultCount;
  final String? category;
  final List<String> filters;

  SearchHistoryItem({
    required this.id,
    required this.query,
    required this.timestamp,
    this.resultCount = 0,
    this.category,
    this.filters = const [],
  });
}

class SearchHistoryScreen extends StatefulWidget {
  const SearchHistoryScreen({super.key});

  @override
  State<SearchHistoryScreen> createState() => _SearchHistoryScreenState();
}

class _SearchHistoryScreenState extends State<SearchHistoryScreen> {
  List<SearchHistoryItem> _searchHistory = [];
  List<SearchHistoryItem> _filteredHistory = [];
  final TextEditingController _searchController = TextEditingController();
  bool _isLoading = true;
  String _selectedTimeFilter = 'all';
  bool _isSelectionMode = false;
  final Set<String> _selectedItems = {};

  @override
  void initState() {
    super.initState();
    _loadSearchHistory();
    _searchController.addListener(_filterHistory);
  }

  @override
  void dispose() {
    _searchController.dispose();
    super.dispose();
  }

  Future<void> _loadSearchHistory() async {
    setState(() => _isLoading = true);

    // Simulate loading from local storage/database
    await Future.delayed(const Duration(milliseconds: 500));

    final mockHistory = _generateMockHistory();

    setState(() {
      _searchHistory = mockHistory;
      _filteredHistory = mockHistory;
      _isLoading = false;
    });
  }

  List<SearchHistoryItem> _generateMockHistory() {
    final queries = [
      'Flutter tutorials',
      'Dart programming',
      'Mobile app development',
      'UI/UX design',
      'State management',
      'API integration',
      'Database design',
      'Authentication',
      'Push notifications',
      'App deployment',
      'Performance optimization',
      'Widget testing',
      'Clean architecture',
      'Responsive design',
      'Dark mode implementation',
    ];

    final categories = ['Development', 'Design', 'Architecture', 'Testing'];

    return List.generate(25, (index) {
      final daysAgo = index * 2 + (index % 3);
      return SearchHistoryItem(
        id: 'search_$index',
        query: queries[index % queries.length],
        timestamp: DateTime.now().subtract(Duration(days: daysAgo)),
        resultCount: 50 + (index * 15),
        category: categories[index % categories.length],
        filters: index % 3 == 0
            ? ['Electronics', 'Popular']
            : index % 4 == 0
                ? ['Recent', 'Trending']
                : [],
      );
    });
  }

  void _filterHistory() {
    final query = _searchController.text.toLowerCase();
    setState(() {
      _filteredHistory = _searchHistory.where((item) {
        final matchesSearch = item.query.toLowerCase().contains(query);
        final matchesTimeFilter = _matchesTimeFilter(item);
        return matchesSearch && matchesTimeFilter;
      }).toList();
    });
  }

  bool _matchesTimeFilter(SearchHistoryItem item) {
    final now = DateTime.now();
    switch (_selectedTimeFilter) {
      case 'today':
        return item.timestamp.isAfter(DateTime(now.year, now.month, now.day));
      case 'week':
        return item.timestamp.isAfter(now.subtract(const Duration(days: 7)));
      case 'month':
        return item.timestamp.isAfter(now.subtract(const Duration(days: 30)));
      default:
        return true;
    }
  }

  void _performSearch(String query) {
    Navigator.pushNamed(
      context,
      '/search-results',
      arguments: {'query': query},
    );
  }

  void _deleteHistoryItem(String id) {
    setState(() {
      _searchHistory.removeWhere((item) => item.id == id);
      _filteredHistory.removeWhere((item) => item.id == id);
    });
    // In a real app, you would also remove from persistent storage
    _showSnackBar('Search removed from history');
  }

  void _deleteSelectedItems() {
    setState(() {
      _searchHistory.removeWhere((item) => _selectedItems.contains(item.id));
      _filteredHistory.removeWhere((item) => _selectedItems.contains(item.id));
      _selectedItems.clear();
      _isSelectionMode = false;
    });
    _showSnackBar('${_selectedItems.length} items removed');
  }

  void _clearAllHistory() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Clear Search History'),
        content: const Text('Are you sure you want to clear all search history? This action cannot be undone.'),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: const Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              setState(() {
                _searchHistory.clear();
                _filteredHistory.clear();
              });
              Navigator.pop(context);
              _showSnackBar('Search history cleared');
            },
            child: const Text('Clear All'),
          ),
        ],
      ),
    );
  }

  void _showSnackBar(String message) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(message)),
    );
  }

  void _toggleSelectionMode() {
    setState(() {
      _isSelectionMode = !_isSelectionMode;
      if (!_isSelectionMode) {
        _selectedItems.clear();
      }
    });
  }

  void _toggleItemSelection(String id) {
    setState(() {
      if (_selectedItems.contains(id)) {
        _selectedItems.remove(id);
      } else {
        _selectedItems.add(id);
      }
    });
  }

  String _getRelativeTime(DateTime timestamp) {
    final now = DateTime.now();
    final difference = now.difference(timestamp);

    if (difference.inDays > 7) {
      return '${timestamp.day}/${timestamp.month}/${timestamp.year}';
    } else if (difference.inDays > 0) {
      return '${difference.inDays} day${difference.inDays > 1 ? 's' : ''} ago';
    } else if (difference.inHours > 0) {
      return '${difference.inHours} hour${difference.inHours > 1 ? 's' : ''} ago';
    } else if (difference.inMinutes > 0) {
      return '${difference.inMinutes} minute${difference.inMinutes > 1 ? 's' : ''} ago';
    } else {
      return 'Just now';
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: _isSelectionMode
            ? Text('${_selectedItems.length} selected')
            : const Text('Search History'),
        leading: _isSelectionMode
            ? IconButton(
                icon: const Icon(Icons.close),
                onPressed: _toggleSelectionMode,
              )
            : null,
        actions: [
          if (_isSelectionMode) ...[
            if (_selectedItems.isNotEmpty)
              IconButton(
                icon: const Icon(Icons.delete),
                onPressed: _deleteSelectedItems,
              ),
          ] else ...[
            IconButton(
              icon: const Icon(Icons.select_all),
              onPressed: _toggleSelectionMode,
              tooltip: 'Select items',
            ),
            PopupMenuButton<String>(
              onSelected: (value) {
                switch (value) {
                  case 'clear_all':
                    _clearAllHistory();
                    break;
                }
              },
              itemBuilder: (context) => [
                const PopupMenuItem(
                  value: 'clear_all',
                  child: Row(
                    children: [
                      Icon(Icons.clear_all, size: 20),
                      SizedBox(width: 8),
                      Text('Clear All History'),
                    ],
                  ),
                ),
              ],
            ),
          ],
        ],
      ),
      body: Column(
        children: [
          // Search Bar
          Container(
            padding: const EdgeInsets.all(16),
            child: TextField(
              controller: _searchController,
              decoration: InputDecoration(
                hintText: 'Search in history...',
                prefixIcon: const Icon(Icons.search),
                suffixIcon: _searchController.text.isNotEmpty
                    ? IconButton(
                        icon: const Icon(Icons.clear),
                        onPressed: () {
                          _searchController.clear();
                          _filterHistory();
                        },
                      )
                    : null,
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(12),
                ),
                filled: true,
                fillColor: Theme.of(context).colorScheme.surface,
              ),
            ),
          ),

          // Time Filter Chips
          Container(
            height: 50,
            padding: const EdgeInsets.symmetric(horizontal: 16),
            child: ListView(
              scrollDirection: Axis.horizontal,
              children: [
                _buildTimeFilterChip('All time', 'all'),
                _buildTimeFilterChip('Today', 'today'),
                _buildTimeFilterChip('This week', 'week'),
                _buildTimeFilterChip('This month', 'month'),
              ],
            ),
          ),

          // Content
          Expanded(
            child: _isLoading
                ? const Center(child: CircularProgressIndicator())
                : _filteredHistory.isEmpty
                    ? _buildEmptyState()
                    : _buildHistoryList(),
          ),
        ],
      ),
    );
  }

  Widget _buildTimeFilterChip(String label, String value) {
    final isSelected = _selectedTimeFilter == value;
    return Container(
      margin: const EdgeInsets.only(right: 8),
      child: FilterChip(
        label: Text(label),
        selected: isSelected,
        onSelected: (selected) {
          setState(() {
            _selectedTimeFilter = value;
            _filterHistory();
          });
        },
        backgroundColor: Theme.of(context).colorScheme.surface,
        selectedColor: Theme.of(context).colorScheme.primaryContainer,
      ),
    );
  }

  Widget _buildEmptyState() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(
            Icons.history,
            size: 64,
            color: Colors.grey[400],
          ),
          const SizedBox(height: 16),
          Text(
            _searchController.text.isNotEmpty
                ? 'No matching searches found'
                : 'No search history',
            style: Theme.of(context).textTheme.titleLarge,
          ),
          const SizedBox(height: 8),
          Text(
            _searchController.text.isNotEmpty
                ? 'Try different keywords'
                : 'Your recent searches will appear here',
            style: Theme.of(context).textTheme.bodyMedium?.copyWith(
              color: Colors.grey[600],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildHistoryList() {
    // Group by date
    final groupedHistory = <String, List<SearchHistoryItem>>{};
    for (final item in _filteredHistory) {
      final now = DateTime.now();
      final itemDate = item.timestamp;
      String groupKey;

      if (itemDate.isAfter(DateTime(now.year, now.month, now.day))) {
        groupKey = 'Today';
      } else if (itemDate.isAfter(DateTime(now.year, now.month, now.day - 1))) {
        groupKey = 'Yesterday';
      } else if (itemDate.isAfter(now.subtract(const Duration(days: 7)))) {
        groupKey = 'This week';
      } else if (itemDate.isAfter(now.subtract(const Duration(days: 30)))) {
        groupKey = 'This month';
      } else {
        groupKey = 'Older';
      }

      groupedHistory.putIfAbsent(groupKey, () => []).add(item);
    }

    return ListView.builder(
      padding: const EdgeInsets.all(16),
      itemCount: groupedHistory.length * 2, // Headers + items
      itemBuilder: (context, index) {
        final groupKeys = groupedHistory.keys.toList();
        final groupIndex = index ~/ 2;

        if (index % 2 == 0) {
          // Group header
          if (groupIndex >= groupKeys.length) return const SizedBox.shrink();
          final groupKey = groupKeys[groupIndex];
          return _buildGroupHeader(groupKey);
        } else {
          // Group items
          if (groupIndex >= groupKeys.length) return const SizedBox.shrink();
          final groupKey = groupKeys[groupIndex];
          final items = groupedHistory[groupKey]!;
          return Column(
            children: items.map((item) => _buildHistoryItem(item)).toList(),
          );
        }
      },
    );
  }

  Widget _buildGroupHeader(String title) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 24, 16, 8),
      child: Align(
        alignment: Alignment.centerLeft,
        child: Text(
          title,
          style: Theme.of(context).textTheme.titleSmall?.copyWith(
            color: Theme.of(context).primaryColor,
            fontWeight: FontWeight.bold,
          ),
        ),
      ),
    );
  }

  Widget _buildHistoryItem(SearchHistoryItem item) {
    final isSelected = _selectedItems.contains(item.id);

    return Card(
      margin: const EdgeInsets.only(bottom: 8),
      child: ListTile(
        leading: _isSelectionMode
            ? Checkbox(
                value: isSelected,
                onChanged: (_) => _toggleItemSelection(item.id),
              )
            : CircleAvatar(
                backgroundColor: Theme.of(context).primaryColor.withOpacity(0.1),
                child: Icon(
                  Icons.history,
                  color: Theme.of(context).primaryColor,
                ),
              ),
        title: Text(
          item.query,
          style: const TextStyle(fontWeight: FontWeight.w500),
        ),
        subtitle: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(_getRelativeTime(item.timestamp)),
            if (item.category != null || item.filters.isNotEmpty) ...[
              const SizedBox(height: 4),
              Wrap(
                spacing: 4,
                children: [
                  if (item.category != null)
                    Container(
                      padding: const EdgeInsets.symmetric(
                        horizontal: 6,
                        vertical: 2,
                      ),
                      decoration: BoxDecoration(
                        color: Theme.of(context).primaryColor.withOpacity(0.1),
                        borderRadius: BorderRadius.circular(4),
                      ),
                      child: Text(
                        item.category!,
                        style: Theme.of(context).textTheme.bodySmall,
                      ),
                    ),
                  ...item.filters.map((filter) => Container(
                    padding: const EdgeInsets.symmetric(
                      horizontal: 6,
                      vertical: 2,
                    ),
                    decoration: BoxDecoration(
                      color: Colors.grey.withOpacity(0.2),
                      borderRadius: BorderRadius.circular(4),
                    ),
                    child: Text(
                      filter,
                      style: Theme.of(context).textTheme.bodySmall,
                    ),
                  )),
                ],
              ),
            ],
          ],
        ),
        trailing: _isSelectionMode
            ? null
            : PopupMenuButton<String>(
                onSelected: (value) {
                  switch (value) {
                    case 'copy':
                      Clipboard.setData(ClipboardData(text: item.query));
                      _showSnackBar('Query copied to clipboard');
                      break;
                    case 'delete':
                      _deleteHistoryItem(item.id);
                      break;
                  }
                },
                itemBuilder: (context) => [
                  const PopupMenuItem(
                    value: 'copy',
                    child: Row(
                      children: [
                        Icon(Icons.copy, size: 20),
                        SizedBox(width: 8),
                        Text('Copy'),
                      ],
                    ),
                  ),
                  const PopupMenuItem(
                    value: 'delete',
                    child: Row(
                      children: [
                        Icon(Icons.delete, size: 20),
                        SizedBox(width: 8),
                        Text('Delete'),
                      ],
                    ),
                  ),
                ],
              ),
        onTap: _isSelectionMode
            ? () => _toggleItemSelection(item.id)
            : () => _performSearch(item.query),
        selected: isSelected,
      ),
    );
  }
}